extensions/amp-date-display/1.0/web-component.js

```diff
@@ -0,0 +1,31 @@
+import mustache from '#bento/components/bento-mustache/1.0/bento-mustache';
+import {getTemplate} from '#bento/util/template';
+
+import {defineBentoElement} from '#preact/bento-ce';
+
+import {BaseElement, TAG} from './base-element';
+
+export class BentoDateDisplay extends BaseElement {
+  /** @override */
+  checkPropsPostMutations() {
+    const template = getTemplate(this.element);
+    if (!template) {
+      // show error?
+      return;
+    }
+
+    this.mutateProps({
+      'render': (data) => {
+        const html = mustache.render(template./*OK*/ innerHTML, data);
```

Comment:
```suggestion
        const html = mustache.render(template.content.innerHTML, data);
```

The correct way to use the `<template>` tag is to access the template via the `contents` property.  Unfortunately, the `innerHTML` property will be empty.

But surprisingly, you've probably found that this works CORRECTLY in unit tests or Storybook, right?  I recently discovered an issue with Preact and the `<template>` tag.  Since Preact programmatically creates the DOM, it appends all `<template>` children as "real" children, so `innerHTML` works. 
 In an actual web browser, however, the HTML parser behaves differently!  The HTML parser for `<template>` will create a document fragment (`template.content`) and append all children there instead.  The `<template>` tag will have NO children of its own!  And therefore, `innerHTML` will be empty.

Reply:
So I propose the following solution:
1. Make the above change, using `template.content.innerHTML` as suggested.
2. Here is a way to workaround the Preact bug:
```
        <template
          dangerouslySetInnerHTML={{
            __html: `
              <div>
                <h3>Name: {{name}} ({{username}})</h3>
                <p>Contact: {{email}}; {{phone}}</p>
              </div>
            `,
          }}
        />
```

Reply:
My only caveat is this:  unfortunately I have NOT found any documentation regarding this Preact issue.  I have not yet filed an issue.  I'm not 100% confident that this is a widespread issue; it's reasonable that this issue only exists in our project.  
But I have certainly verified the behavior, and our own codebase supports using `template.content`: https://github.com/ampproject/amphtml/blob/main/src/core/dom/index.js#L303-L311

Reply:
Thanks for detailed explanation @scottrippey. Couple of things to note:
1. `template.innerHTML` does work (although I agree there isn't much documentation about it). See https://jsbin.com/koziriw/edit?html,js,output - I verified on Chrome, Safari & FF
2. `template.content` returns a DocumentFragment, which does not have an `innerHTML` property. Since `mustache.render` requires a string as the first param, one way is to do this is iterate through child nodes and build it. TBH not a fan of this approach.

I'll let @jridgewell chime in here.

Reply:
Hi all,

`<template>` is a weird element, and you're both partially correct. When the HTML parser parses a `<template>` element, it creates a new `DocumentFragment`, and appends all parse children into the fragment. This is so the children are inert and not an active part of the DOM tree. Preact also does not construct `<template>` element correctly `<template><div /></template>` will insert `<div>` as a child of the template (as in `firstChild` like any other element) instead of as a child of the fragment.

But these two points shouldn't affect our usage here. The template @dmanek is querying and reading from is will be parsed out of HTML bytes, so its children will correctly exist inside the fragment. But, the `innerHTML` method is overridden for `<template>`s, and it actually stringifies from the fragment's instead of the `<template>`'s (which shouldn't exist). This works even though `DocumentFragment` doesn't actually have an `innerHTML` getter.

This falls out of https://html.spec.whatwg.org/multipage/scripting.html#the-template-element:
> The [template contents](https://html.spec.whatwg.org/multipage/scripting.html#template-contents) of a [template](https://html.spec.whatwg.org/multipage/scripting.html#the-template-element) element [are not children of the element itself](https://html.spec.whatwg.org/multipage/syntax.html#template-syntax).

If we then look at the `innerHTML` spec, it specifically calls out `<template>` elements: https://w3c.github.io/DOM-Parsing/#dom-innerhtml-innerhtml
> If the [context object](https://w3c.github.io/DOM-Parsing/#dfn-context-object) is a [template](https://w3c.github.io/DOM-Parsing/#dfn-template) element, then let [context object](https://w3c.github.io/DOM-Parsing/#dfn-context-object) be the [template](https://w3c.github.io/DOM-Parsing/#dfn-template)'s [template contents](https://w3c.github.io/DOM-Parsing/#dfn-template-content) (a [DocumentFragment](https://w3c.github.io/DOM-Parsing/#dfn-documentfragment)).

Reply:
I looked more carefully, and yes, you're right, the `template.innerHTML` is definitely working as you said, and `template.content.innerHTML` is not valid.  So no changes need to be made.

In the `amp-mustache.js` implementation, there's a lot of extra steps involved in extracting the `innerHTML` -- cloning the `template.content` and then appending the clone into an empty `div` to extract the `innerHTML` ... it all seems unnecessary now, knowing that `template.innerHTML` works correctly!

So I guess the usefulness of this comment thread is just to point out that Preact still has a bug, where if you did `<template><span>hello</span></template>` in a Preact-based unit test or Storybook, the template would have `innerHTML === ""`.  The workaround is to use `<template dangerouslySetInnerHTML=... />`.

Reply:
Thanks for both for your input. Will resolve this since there is nothing to do on my part.